Dynamically allocate `struct mount` paths

dinhngtu opened the pull request:
### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Description of changes

Currently `m_path` and `m_special` are statically-allocated arrays, meaning their storage spaces are wasted in `static struct mount` variables.

Dynamically allocate these variables instead to save on binary size.

This PR saves ~8K of uncompressed binary size per `static struct mount` object.

dinhngtu committed:
lib/vfscore: Dynamically allocate mount paths

Currently `m_path` and `m_special` are statically-allocated arrays,
meaning their storage spaces are wasted in `static struct mount`
variables.  Dynamically allocate these variables to save on binary size.

Also add initcalls to initialize `struct mounts` in lib/posix-event,
lib/posix-socket and lib/vfscore.

Signed-off-by: Tu Dinh Ngoc <dinhngoc.tu@irit.fr>

RaduNichita reviewed (approved):
I tested with `app-nginx`, `app-redis` and `app-sqlite` on x86 and everything was OK. 

Thanks @dinhngtu  for providing this PR.

Reviewed-by:  Radu Nichita <radunichita99@gmail.com>

skuenzer commented on lib/vfscore/mount.c:
```diff
@@ -171,8 +171,8 @@ UK_SYSCALL_R_DEFINE(int, mount, const char*, dev, const char*, dir,
 	mp->m_data = NULL;
-	strlcpy(mp->m_path, dir, sizeof(mp->m_path));
-	strlcpy(mp->m_special, dev, sizeof(mp->m_special));
+	mp->m_path = strndup(dir, PATH_MAX - 1);
```
Can you also update the code to abort operation if the allocation failed (`ENOMEM`)?

skuenzer commented on lib/vfscore/mount.c:
```diff
@@ -236,6 +244,8 @@ UK_SYSCALL_R_DEFINE(int, mount, const char*, dev, const char*, dir,
+	free(mp->m_special);
+	free(mp->m_path);
 	free(mp);
  err1:
```
I'd actually rename this label to `err_close_device:`

skuenzer commented on lib/vfscore/mount.c:
```diff
@@ -236,6 +244,8 @@ UK_SYSCALL_R_DEFINE(int, mount, const char*, dev, const char*, dir,
  err2:
+	free(mp->m_special);
+	free(mp->m_path);
 	free(mp);
```
`err_free_mp:`

skuenzer commented on lib/vfscore/mount.c:
```diff
@@ -236,6 +244,8 @@ UK_SYSCALL_R_DEFINE(int, mount, const char*, dev, const char*, dir,
 		drele(dp_covered);
  err2:
+	free(mp->m_special);
+	free(mp->m_path);
```
`err_free_mp_m_path:`

skuenzer commented on lib/vfscore/mount.c:
```diff
@@ -236,6 +244,8 @@ UK_SYSCALL_R_DEFINE(int, mount, const char*, dev, const char*, dir,
 	if (dp_covered)
 		drele(dp_covered);
  err2:
+	free(mp->m_special);
```
If a jump needs to go here: `err_free_mp_m_special:`

skuenzer commented on lib/vfscore/mount.c:
```diff
@@ -171,8 +171,16 @@ UK_SYSCALL_R_DEFINE(int, mount, const char*, dev, const char*, dir,
+	mp->m_path = strndup(dir, PATH_MAX - 1);
+	if (!mp->m_path) {
+		error = ENOMEM;
+		goto err2;
```
On errors `mp->m_path` and `mp->m_special` are empty, so you should go to free'ing `mp`, instead.
```c
goto err_free_mp;
```

skuenzer commented on lib/vfscore/mount.c:
```diff
@@ -171,8 +171,16 @@ UK_SYSCALL_R_DEFINE(int, mount, const char*, dev, const char*, dir,
+	mp->m_special = strndup(dev, PATH_MAX - 1);
+	if (!mp->m_special) {
+		error = ENOMEM;
+		goto err2;
```
```c
goto err_free_m_path;
```

skuenzer reviewed (changes requested):
Hi @dinhngtu,

thanks a lot for the revision. Can you update the jump labels? Normally we avoid doing `free(NULL)`, so you would need to split your `err2:` block.

Thanks,

Simon

dinhngtu commented:
I have updated the jump labels as requested.

nderjung reviewed (changes requested):
Thanks for this. Just a note on the Git title formats. The two commit titles should be:
```
lib: Add initcalls to initialize `struct mounts`
lib/vfscore: Dynamically allocate struct mount paths
```
Thanks!

dinhngtu commented:
I've renamed them. The first commit message was shortened slightly to fit into the length limit.

skuenzer commented on lib/posix-event/epoll.c:
```diff
@@ -371,3 +373,17 @@ int epoll_pwait(int epfd, struct epoll_event *events, int maxevents,
+		return -ENOMEM;
+	epoll_mount.m_special = strdup("");
+	if (!epoll_mount.m_special) {
+		free(epoll_mount.m_path);
```
I'd actually prefer a goto logic as we do in general for error exits. The reason is that this can be extended more easily later:

```c
static int epoll_mount_init(void)
{
	epoll_mount.m_path = strdup("");
	if (!epoll_mount.m_path)
		goto err_out;
	epoll_mount.m_special = strdup("");
	if (!epoll_mount.m_special) {
		git err_free_m_path;
	return 0;

err_free_m_path:
	free(epoll_mount.m_path);
err_out:
	return -ENOMEM;
}
```

skuenzer commented on lib/posix-event/eventfd.c:
```diff
@@ -438,3 +440,17 @@ int eventfd(unsigned int initval, int flags)
+		return -ENOMEM;
+	eventfd_mount.m_special = strdup("");
+	if (!eventfd_mount.m_special) {
+		free(eventfd_mount.m_path);
```
same here

skuenzer commented on lib/posix-socket/socket_vnops.c:
```diff
@@ -342,3 +344,17 @@ static struct vfsops posix_socket_vfsops = {
+		return -ENOMEM;
+	posix_socket_mount.m_special = strdup("");
+	if (!posix_socket_mount.m_special) {
+		free(posix_socket_mount.m_path);
```
same here

skuenzer commented on lib/vfscore/pipe.c:
```diff
@@ -718,3 +719,17 @@ int mkfifo(const char *path __unused, mode_t mode __unused)
+		return -ENOMEM;
+	p_mount.m_special = strdup("");
+	if (!p_mount.m_special) {
+		free(p_mount.m_path);
```
same here

skuenzer reviewed (commented):
Hi @dinhngtu,

thanks for the update. I'd still prefer a different prefix for the second commit, something like `lib/posix-event, lib/posix-socket, lib/vfscore:`. Unfortunately, this is pretty long.
However, I noticed that this change is related to the first commit. Probably you should just merge them together into a single commit and describe in the commit subject that you allocate empty strings for mount structs for lib/posix-event, lib/posix-socket, lib/vfscore, too. Btw, `NULL` or just `""` instead of `strdup("")` would not work? Are these mount structs ever free'd?

Thanks,

Simon

dinhngtu commented:
I've changed the init logic accordingly and squashed the commits.

I didn't want to use `NULL` or `""` for the mount paths in these structs since I wanted to kept these strings being dynamically-allocated strings like other mount structs. I think it'd be confusing to have some mount path strings being dynamically allocated and others not.

skuenzer commented:
> I've changed the init logic accordingly and squashed the commits.
> 
> I didn't want to use `NULL` or `""` for the mount paths in these structs since I wanted to kept these strings being dynamically-allocated strings like other mount structs. I think it'd be confusing to have some mount path strings being dynamically allocated and others not.

I agree with `""` but `NULL` would be still okay, however, I am not sure if a lot of code needs to be patched for this case.
Other from this the changes are fine to me.

skuenzer reviewed (approved):
Approved-by: Simon Kuenzer <simon@unikraft.io>
